src/python/pants/backend/cc/util_rules/toolchain.py

```diff
@@ -0,0 +1,212 @@
+            compiler=subsystem.cpp_executable,
+            compile_flags=tuple(subsystem.cpp_compile_options),
+            compile_defines=tuple(subsystem.cpp_defines),
+            digest=extracted_archive.digest,
+        )
+
+    return CCToolchain(
+        compiler=subsystem.c_executable,
+        compile_flags=tuple(subsystem.c_compile_options),
+        compile_defines=tuple(subsystem.c_defines),
+        digest=extracted_archive.digest,
+    )
+
+
+@rule_helper
+async def _setup_system_toolchain(
+    request: CCToolchainRequest, subsystem: CCSubsystem
+) -> CCToolchain:
+    """Set up a toolchain from the user's host system."""
+
+    # Sanitize the search paths in case the "<PATH>" is specified
+    raw_search_paths = list(subsystem.search_paths)
+    if "<PATH>" in raw_search_paths:
+        i = raw_search_paths.index("<PATH>")
+        env = await Get(Environment, EnvironmentRequest(["PATH"]))
+        system_path = env.get("PATH", "")
+        raw_search_paths[i : i + 1] = system_path.split(os.pathsep)
+
+    search_paths = tuple(OrderedSet(raw_search_paths))
+
+    # Populate the toolchain for C or C++ accordingly
+    if request.language == CCLanguage.CPP:
+        cpp_executable = await _executable_path(tuple(subsystem.cpp_executable), search_paths)
+        return CCToolchain(
+            cpp_executable,
+            compile_flags=tuple(subsystem.cpp_compile_options),
+            compile_defines=tuple(subsystem.cpp_defines),
+        )
+
+    c_executable = await _executable_path(tuple(subsystem.c_executable), search_paths)
+    return CCToolchain(
+        c_executable,
+        compile_flags=tuple(subsystem.c_compile_options),
+        compile_defines=tuple(subsystem.c_defines),
+    )
+
+
+@rule(desc="Setup the CC Toolchain", level=LogLevel.DEBUG)
+async def setup_cc_toolchain(
+    request: CCToolchainRequest,
+    subsystem: CCSubsystem,
+    external_subsystem: ExternalCCSubsystem,
+    platform: Platform,
+) -> CCToolchain:
+    """Set up the C/C++ toolchain."""
+    if external_subsystem.url_template:
+        return await _setup_downloadable_toolchain(request, external_subsystem, platform)
+    else:
+        return await _setup_system_toolchain(request, subsystem)
+
+
+@dataclass(frozen=True)
+class CCProcess:
+    args: tuple[str, ...]
+    language: CCLanguage
+    description: str
+    input_digest: Digest = EMPTY_DIGEST
+    output_files: tuple[str, ...] = ()
+    level: LogLevel = LogLevel.INFO
+
+
+@rule(desc="Setup a CC Process loaded with the CCToolchain", level=LogLevel.DEBUG)
+async def setup_cc_process(request: CCProcess) -> Process:
+    """Set up a C/C++ process.
+
+    This rule will load the C/C++ toolchain based on the requested language. It will then return a
+    Process that can be run to compile or link a C/C++ source file.
+    """
+    toolchain = await Get(CCToolchain, CCToolchainRequest(request.language))
+
+    # TODO: What if this is for linking instead of compiling?
```

Comment:
Should there then be a `CCCompilerProcess` and `CCLinkerProcess`? (Probably a bit much to split like that but good for me to at least ask the question.)

Reply:
Hmmmm, yeah, that sounds like a good abstraction to add. I'll test it out and see how it feels vs alternatives.